### PR TITLE
Honor configs & report public hostname on status

### PR DESCRIPTION
**Pull request by benjaminapetersen**
- handles values set on console config
- handle values set on operator config
- report public hostname on status
   - oauth team dependent on this update

### CODE REVIEW

**Comment by benjaminapetersen** [comment]
/hold

Needs some work to guard/check/validate/etc.

@jhadvig @zherman0 fyi, we will need this in this week.

File: pkg/console/subresource/configmap/configmap.go [hunk 1]
```diff
@@ -18,23 +19,57 @@ const (
...
+const (
+	defaultLogoutURL = ""
+)
+
+// overridden by operator config
+const (
+	defaultDocumentationBaseURL = "https://docs.okd.io/4.0/"
+	defaultBranding             = "okd"
+)
+
```
Comment (benjaminapetersen) [inline pkg/console/subresource/configmap/configmap.go]: These 3 functions may belong in `util.go` or somewhere else...

File: pkg/console/subresource/configmap/configmap.go [hunk 2]
```diff
@@ -18,23 +19,57 @@ const (
...
+//consoleConfig.Spec.Authentication.LogoutRedirect
+func getLogoutRedirect(consoleConfig *configv1.Console) string {
+	// TODO: handle if not provided
+	// return consoleConfig.Spec.Authentication.LogoutRedirect || defaultLogoutURL
+	return consoleConfig.Spec.Authentication.LogoutRedirect
+}
+
+//operatorConfig.Spec.Customization.Brand
+func getBrand(operatorConfig *operatorv1.Console) operatorv1.Brand {
+	// TODO: handle if not provided
```
Comment (benjaminapetersen) [inline pkg/console/subresource/configmap/configmap.go]: @zherman0 this is related to your brand work PR.

Comment (benjaminapetersen) [reply pkg/console/subresource/configmap/configmap.go]: #126 
Just want to ensure they function together.

Comment (benjaminapetersen) [reply pkg/console/subresource/configmap/configmap.go]: So, if brand is not EXPLICITLY set, it should just fallback to the default which is provided in #126.

File: pkg/console/subresource/configmap/configmap.go [hunk 3]
```diff
@@ -18,23 +19,57 @@ const (
...
-	brandingDefault      = "okd"
 	// serving info
 	certFilePath = "/var/serving-cert/tls.crt"
 	keyFilePath  = "/var/serving-cert/tls.key"
 )
 
-func DefaultConfigMap(cr *operatorv1.Console, rt *routev1.Route) *corev1.ConfigMap {
+// overridden by console config
+const (
+	defaultLogoutURL = ""
```
Comment (jhadvig) [inline pkg/console/subresource/configmap/configmap.go]: dont we have a default URL for logout ?

File: pkg/console/subresource/configmap/configmap.go [hunk 4]
```diff
@@ -18,23 +19,57 @@ const (
...
+// overridden by operator config
+const (
+	defaultDocumentationBaseURL = "https://docs.okd.io/4.0/"
+	defaultBranding             = "okd"
+)
+
+//consoleConfig.Spec.Authentication.LogoutRedirect
+func getLogoutRedirect(consoleConfig *configv1.Console) string {
+	// TODO: handle if not provided
+	// return consoleConfig.Spec.Authentication.LogoutRedirect || defaultLogoutURL
```
Comment (jhadvig) [inline pkg/console/subresource/configmap/configmap.go]: Lets uncomment this and comment the below ? or am I missing something ?

File: pkg/console/subresource/configmap/configmap.go (hunk 3 above)
Comment (benjaminapetersen) [reply pkg/console/subresource/configmap/configmap.go]: Nope, there should not be a default (empty string).  See console auth/logout handling:

https://github.com/openshift/console/blob/e267fcf4e177da5858c5c63d021d4f19f71646b8/frontend/public/module/auth.js#L56

File: pkg/console/subresource/configmap/configmap.go (hunk 4 above)
Comment (benjaminapetersen) [reply pkg/console/subresource/configmap/configmap.go]: Nope, you aren't missing anything. This is why its still WIP.  I didn't have time to test everything before opening last night, but wanted to raise awareness early.

**Comment by jhadvig** [comment]
We probrably need to update the tests
```
[INFO] Testing github.com/openshift/console-operator/pkg/console/subresource/configmap
# github.com/openshift/console-operator/pkg/console/subresource/configmap
pkg/console/subresource/configmap/configmap_test.go:100:30: not enough arguments in call to DefaultConfigMap
	have (*"github.com/openshift/console-operator/vendor/github.com/openshift/api/operator/v1".Console, *"github.com/openshift/console-operator/vendor/github.com/openshift/api/route/v1".Route)
	want (*"github.com/openshift/console-operator/vendor/github.com/openshift/api/operator/v1".Console, *"github.com/openshift/console-operator/vendor/github.com/openshift/api/config/v1".Console, *"github.com/openshift/console-operator/vendor/github.com/openshift/api/route/v1".Route)
pkg/console/subresource/configmap/configmap_test.go:170:14: undefined: NewYamlConfigString
FAIL	github.com/openshift/console-operator/pkg/console/subresource/configmap [build failed]
```

File: pkg/console/subresource/configmap/configmap.go [hunk 5]
```diff
@@ -18,23 +19,57 @@ const (
...
+)
+
+// overridden by operator config
+const (
+	defaultDocumentationBaseURL = "https://docs.okd.io/4.0/"
+	defaultBranding             = "okd"
+)
+
+func getLogoutRedirect(consoleConfig *configv1.Console) string {
+	if len(consoleConfig.Spec.Authentication.LogoutRedirect) > 0 {
```
Comment (benjaminapetersen) [inline pkg/console/subresource/configmap/configmap.go]: I believe this is all we need to do.  The `config` structs are not pointers, so when the YAML is instantiated, nested structs should be created with empty string values.  Therefore if empty string is provided, we can return the default value instead.

@jhadvig @zherman0 work?  Same treatment for all three of these.

File: pkg/console/subresource/configmap/configmap.go [hunk 6]
```diff
@@ -18,23 +19,57 @@ const (
...
+
+func DefaultConfigMap(operatorConfig *operatorv1.Console, consoleConfig *configv1.Console, rt *routev1.Route) *corev1.ConfigMap {
+
+	logoutRedirect := getLogoutRedirect(consoleConfig)
+	brand := getBrand(operatorConfig)
+	docURL := getDocURL(operatorConfig)
+
 	host := rt.Spec.Host
-	config := NewYamlConfigString(host)
+	config := string(NewYamlConfig(host, logoutRedirect, brand, docURL))
```
Comment (benjaminapetersen) [inline pkg/console/subresource/configmap/configmap.go]: I eliminated the one wrapper func `NewYamlConfigString()`, was unnecessary additional plumbing for passing these values down, when all it did was call `string()`.

**Comment by benjaminapetersen** [comment]
Updated the tests & filled out the defaults.

**Comment by benjaminapetersen** [comment]
/retest

one success. interesting.

**Comment by benjaminapetersen** [comment]
So at present you should be able to update the `console.config.openshift.io`:
```
apiVersion: config.openshift.io/v1
kind: Console
spec:
  authentication:
    logoutRedirect: https://www.redhat.com/i/should/logout/i/think
```
and it will update the web console.

In addition, when the `Route.Host` is provided, this will be reported on the `console.config.openshift.io`:

```
oc get console.config.openshift.io console -o yaml
apiVersion: config.openshift.io/v1
kind: Console
metadata:
  name: console
  resourceVersion: "1542289"
  selfLink: /apis/config.openshift.io/v1/consoles/console
  uid: 9a327206-2fce-11e9-92db-021245e32db2
spec:
  authentication:
    logoutRedirect: https://www.redhat.com/i/should/logout/i/think
status:
  publicHostname: https://console-openshift-console.apps.betersen.devcluster.openshift.com
```
Also, the `console.operator.openshift.io` can be configured with:

```
oc get console.operator.openshift.io console -o yaml
apiVersion: operator.openshift.io/v1
kind: Console
metadata:
  name: console
  resourceVersion: "1483224"  
spec:
  customization:
    brand: online
    documentationBaseURL: https://www.redhat.com/someplace/for/docs/give/it/a/try/
  logLevel: ""
  managementState: Managed
  observedConfig: null
  operandSpecs: null
  unsupportedConfigOverrides: null
```
The customizations on branding & documentationBaseURL will be set on the console `configmap`, and will be updated & visible in the console.

**Comment by benjaminapetersen** [comment]
/retest

flake

**Comment by jhadvig** [comment]
Last failure:
```
time="2019-02-13T21:57:19Z" level=info msg="finished syncing operator \"console\" (50.148µs) \n\n"
E0213 21:57:19.063215       1 controller.go:130] {🐼 🐼} failed with: Operation cannot be fulfilled on consoles.config.openshift.io "console": the object has been modified; please apply your changes to the latest version and try again
I0213 21:57:19.063575       1 option.go:62] Console: handling update openshift-console/console: /apis/apps/v1/namespaces/openshift-console/deployments/console
```

**Comment by benjaminapetersen** [comment]
That fail is interesting. Will check in a short bit.

**Comment by jhadvig** [comment]
flake
/test e2e-aws

**Comment by benjaminapetersen** [comment]
all succeed. @jhadvig or @zherman0 just need a lgtm 👍

**Pull request merged**